runtime/ejs-typedarrays.c

```diff
@@ -537,7 +586,95 @@ EJS_DATA_VIEW_METHOD_IMPL(Float64, double, 8);
+     /* 11. Let beginIndex be min(srcLength, max(0, beginInt)). */      \
+     uint32_t beginIndex = min(srcLength, max(0, beginInt));             \
+                                                                        \
+     /* 12. If end is undefined, then let end be srcLength. */          \
+     if (EJSVAL_IS_UNDEFINED(end))                                      \
+        end = NUMBER_TO_EJSVAL(srcLength);                              \
+                                                                        \
+     /* 13. Let endInt be ToInteger(end). */                            \
+     /* 14. ReturnIfAbrupt(endInt). */                                  \
+     int endInt = ToInteger(end);                                       \
+                                                                        \
+     /* 15. If endInt < 0, then let endInt be srcLength + endInt. */    \
+     if (endInt < 0)                                                    \
+        endInt = srcLength + endInt;                                    \
+                                                                        \
+     /* 16. Let endIndex be max(0,min(srcLength, endInt)). */           \
+     uint32_t endIndex = max(0, min(srcLength, endInt));                 \
+                                                                        \
+     /* 17. If endIndex < beginIndex, then let endIndex be beginIndex. */ \
+     if (endIndex < beginIndex)                                         \
+        endIndex = beginIndex;                                          \
+                                                                        \
+     /* 18. Let newLength be endIndex - beginIndex. */                  \
+     uint32_t newLength = endIndex - beginIndex;                        \
+                                                                        \
+     /* 19. Let constructorName be the string value of O’s [[TypedArrayName]] internal slot. */ \
+     /* 20. Let elementType be the string value of the Element Type value in Table 46 for constructorName. */ \
+     /* 21. Let elementSize be the Number value of the Element Size value specified in Table 46 for constructorName. */ \
+     uint32_t elementSize = elementSizeInBytes;                         \
+                                                                        \
+     /* 22. Let srcByteOffset be the value of O’s [[ByteOffset]] internal slot. */ \
+     uint32_t srcByteOffset = Oobj->byteOffset;                         \
+                                                                        \
+     /* 23. Let beginByteOffset be srcByteOffset + beginIndex × elementSize. */ \
+     uint32_t beginByteOffset = srcByteOffset + beginIndex * elementSize; \
+                                                                        \
+     /* 24. Let defaultConstructor be the intrinsic object listed in column one of Table 46 for constructorName. */ \
```

Comment:
from this line to the end of the method, the type we're going to construct is hardcoded (using the cpp macros) to be the same as the array whose implementation we're running.  This won't work for JS subclasses of typed arrays.  The SpeciesConstructor stuff doesn't exist yet (we use the old `@@create` model still), so maybe put a comment here and file an issue saying "%TypedArray%.prototype.subarray doesn't create instances of subclasses".